enzyme/Enzyme/AdjointGenerator.h

```diff
@@ -8440,14 +8526,19 @@ class AdjointGenerator
           getForwardBuilder(Builder2);
           auto origfree = orig->getArgOperand(0);
           auto tofree = gutils->invertPointerM(origfree, Builder2);
-          if (tofree != origfree) {
-            SmallVector<Value *, 2> args = {tofree};
-            CallInst *CI = Builder2.CreateCall(orig->getFunctionType(),
-                                               orig->getCalledFunction(), args);
-            CI->setAttributes(orig->getAttributes());
-          }
+
+          auto rule = [&](Value *tofree) {
+            if (tofree != origfree) {
```

Comment:
Have the whole rule / applyChain inside the if?

Reply:
Also the more I think about this code, I think this would be better as a runtime rather than compile time check. In essence the issue here is that we permit the primal version of a value to be used in a shadow (e.g. stored into shadow data structures/etc). This is necessary to maintain some of the same structure as the original code. However, this may result in multiple copies (and in the case of vector mode width extra many copies). A value check helps but does not completel resolve the problem since the LLVM values may still be distinct even if the runtime data is the same from this.

Reply:
fixed